### ISSUE TITLE
I have processed this model using same feature and different edgelist, having some problems for the result.

### ISSUE THREAD

**Issue by Crystalsudo** [cszhangzhen/ANRL#5]
Hi,
If I input edgelist like this:
(weighted and directed)
0 456 3.5229
0 78 3.3552
0 1022 1.6327
0 989 1.3817
0 56 1.1477
……………………
And feature list like this: (I think 0 means ignoring the feature)
1567 5
0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00
0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00
0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00 0.000000000000000000e+00
……………………
Then I get the same result for every node:
0,0.27159807,-0.5618622,0.389931
1,0.27159807,-0.5618622,0.389931
……
So why the different weights seem have no influence on the embedding result?
Look forward to your reply!

**Comment by cszhangzhen** [comment]
It's hard to say why this result happens.

Are the input features all zeros? If so, maybe you can adjust the trade-off parameter between Skip-gram and Auto-encoder modules to focus more on the available graph structure information.

Moreover, when applying random walk based model to directed graph, I find that the walker usually stops after only a few steps, thus generating very short random walk sequences not sufficient for training the Skip-gram model. You need pay attention to this phenomenon.

**Comment by Crystalsudo** [comment]
Thank you for your reply!
1.
Features are indeed all zeros because I'd like to compare structure proximity and feature proximity preservation using this model. I think the final result shows that features have no influence on them.
According to your advice, I will make alpha and beta smaller because the parameter for skip gram is 1, am i right?
(Actually, I have indeed tested some combinations of alpha and beta, but I will change them more specifically tomorrow.)

2. When I tried different iteration number, I find some strange things:
  If the learning rate is 1e-4, then the model won't converge even after 1,000,000 iterations.
  If the learning rate is 1e-3, it seems overshooting.
  I feel confused about this parameter.
  I observe that most of time, the total loss is decreasing companying by one of its component decreasing while another is increasing.

3. BTW, I want to confirm with you that whether this model optimize skip gram and autoencoder at the same time. Will this order influence the final result?

4. So do you mean you also came across only a few walks when you are employing this model under directed graph?

5. Would you mind testing my situation on your side? I stuck in this problem for a long time. I am sure that I have input edgelist with weight and change the parameter into weighted and directed. (And annotate the label files.)

Look forward to your reply.

**Comment by cszhangzhen** [comment]
Hi,
1. Yes, you can try a smaller alpha to see if it works.

2. May be this is caused by the alternative optimization manner. See 3 for detail. 

3. Actually, the skip-gram model and auto-encoder model are not optimized at the same time. I optimize them alternatively, thus it may cause fluctuation in the training procedure. For each module, I randomly sample a batch of data to optimize it. A more reasonable way is to sample same batch of input X (but with different targets) for each module. You can try this manner. Also, the learning rate for each module can be different.

4. Yes, I think most of random-based models will encounter this problem due to the nature of input graph.

5. Maybe you can send me your dataset and I will test it.

**Comment by Crystalsudo** [comment]
Good morning,
There are four datasets in the zip file.
list.edgelist and list_allzero.feature are the datasets I am using now.
list_denser graph.edgelist is a more denser graph which I reconstruct.
list_meanimputation.feature is the imputation result because some of my nodes have features while others do not. 
The original alpha and beta ratio is 10:1 and I have indeed employed default parameter setting. 

Thank you very much!!

[list data.zip](https://github.com/cszhangzhen/ANRL/files/3454483/list.data.zip)

**Comment by cszhangzhen** [comment]
Hi,

I have run your dataset.

I think the problem is caused by your dataset.

I print out the random sequence and the sequences are almost containing only one or two nodes. Please check the result below:
[343, 1312]
[1176]
[1109]
[377, 226, 543]
[588]
[130, 783]
[1562]
[120, 756]
[36, 528]
[214, 960]
[12, 433]
[24, 456]
[101, 716]
[431]
[87, 433]
[137, 439]
[335, 433]
...
Such random sequences will not help the model to learn good embeddings. You need a new random strategy in directed graph, e.g., random walk with restart.

**Comment by Crystalsudo** [comment]
Yep, I indeed printed out this sequence, too. But my initial idea is that as long as there exist different walk sequences for different nodes, then there will be different embedding results.
In addition, I can get different embedding result using node2vec.

This is the final result using node2vec:

[list_all0.5_using node2vec.emb.zip](https://github.com/cszhangzhen/ANRL/files/3454717/list_all0.5_using.node2vec.emb.zip)

**Comment by cszhangzhen** [comment]
Hi,

The final embedding results in ANRL are inferred through node attributes. So, it will output same results in your case. Once the optimization procedure is finished (e.g., W and b in encoder are fixed), the embedding vectors are inferred (e.g., using W and b) based on attributes (input X). This is indeed a drawback of ANRL, because it does not take structure information into consideration when performing inference. The structure information is only used in model optimization procedure.

I suggest you to try graph convolutional neural network in this case.

**Comment by Crystalsudo** [comment]
Got it, thank you for your kind help!
Have a nice day!
